apis/Google.Cloud.Diagnostics.AspNet/Google.Cloud.Diagnostics.AspNet.Snippets/ErrorReportingSnippets.cs

```diff
@@ -0,0 +1,186 @@
+            {
+                var response = await client.GetAsync($"api/ErrorReporting/{nameof(ErrorReportingController.CatchesAndLogsException)}/{testId}");
+                Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+
+                var errorEvent = _polling.GetEvents(startTime, testId, 1).Single();
+                VerifyFullLogging(errorEvent, testId, nameof(ErrorReportingController.ThrowsException));
+            }
+        }
+
+        internal static void VerifyFullLogging(ErrorEvent errorEvent, string testId, string functionName)
+        {
+            VerifyErrorLogging(errorEvent, testId, functionName);
+            VerifyHttpContextLogging(errorEvent);
+        }
+
+        /// <summary>
+        /// Checks that an <see cref="ErrorEvent"/> contains valid data.
+        /// </summary>
+        /// <param name="errorEvent">The event to check.</param>
+        /// <param name="testId">The id of the test.</param>
+        /// <param name="functionName">The name of the function the error occurred in.</param>
+        internal static void VerifyErrorLogging(ErrorEvent errorEvent, string testId, string functionName)
+        {
+            Assert.Equal(ErrorReportingSnippetsApplication.Service, errorEvent.ServiceContext.Service);
+            Assert.Equal(ErrorReportingSnippetsApplication.Version, errorEvent.ServiceContext.Version);
+
+            Assert.Contains(functionName, errorEvent.Message);
+            Assert.Contains(testId, errorEvent.Message);
+
+            if (Utils.IsWindows())
+            {
+                Assert.False(string.IsNullOrWhiteSpace(errorEvent.Context.ReportLocation.FilePath));
+                Assert.True(errorEvent.Context.ReportLocation.LineNumber > 0);
+            }
+            Assert.Equal(functionName, errorEvent.Context.ReportLocation.FunctionName);
+        }
+
+        /// <summary>
+        /// Only for automatic logging, in self hosted web apis <see cref="HttpContext.Current"/>
+        /// is <code>null</code>.
+        /// </summary>
+        internal static void VerifyHttpContextLogging(ErrorEvent errorEvent)
+        {
+            Assert.Equal(HttpMethod.Get.Method, errorEvent.Context.HttpRequest.Method);
+            Assert.False(string.IsNullOrWhiteSpace(errorEvent.Context.HttpRequest.Url));
+        }
+
+        internal static string GetMessage(string message, string id) => $"{message} - {id}";
+    }
+
+    internal class ErrorReportingSnippetsApplication : HttpApplication
+    {
+        public const string Service = "service-name";
+        public const string Version = "version-id";
+        public static readonly string ProjectId = Utils.GetProjectIdFromEnvironment();
+
+        public void Configuration(IAppBuilder app)
+        {
+            HttpConfiguration config = new HttpConfiguration();
+            Register(config);
+            app.UseWebApi(config);
+        }
+
+        // Sample: RegisterExceptionLoggerWebApi
```

Comment:
This bit *does* look suitable for the snippets, of course.
Did you move the tests to snippets to reduce code duplication? If so, I'd probably live with the duplication, and have a single test that checks the samples, but do most of the testing in the IntegrationTests project. You could have a common testing project if that's useful - we do that in Spanner. Let me know if you to chat about the details.